File: tests/Cpp/Vector2D.cpp
class Vector2D{
public:
    float X;
    float Y;

    Vector2D(float x, float y){
        this->X = x;
        this->Y = y;
    }

    static const Vector2D Zero;

    // Operator overloads
    Vector2D operator+(const Vector2D &v) const {
        return Vector2D(this->X + v.X, this->Y + v.Y);
    }

    Vector2D operator-(const Vector2D &v) const {
        return Vector2D(this->X - v.X, this->Y - v.Y);
    }

    Vector2D operator-() const {
        return Vector2D(-(this->X), -(this->Y));
    }

    Vector2D operator*(float s) const {
        return Vector2D(this->X * s, this->Y * s);
    }

    Vector2D operator/(float s) const {
        return Vector2D(this->X / s, this->Y / s);
    }
};

const Vector2D Vector2D::Zero(0,0);

File: tests/Cs/Vector2D.cs
public struct Vector2D{

  // java "final" = c# "readonly"
  public readonly float X;
  public readonly float Y;

  public static readonly Vector2D Zero = new Vector2D(0,0);

  public Vector2D{
    this.X = x;
    this.Y = y;
  }
  // operators overload.
  public static Vector2D operator +(Vector2D u, Vector2D v){
    return new Vector2D(u.X + v.X, u.Y + v.Y);
  }

  public static Vector2D operator -(Vector2D u, Vector2D v){
    return new Vector2D(u.X - v.X, u.Y - v.Y);
  }

  public static Vector2D operator -(Vector2D u){
    return new Vector2D(-v.X, -v.Y);
  }

  public static Vector2D operator *(Vector2D v, float s){
    return new Vector2D(v.X * s, v.Y * s);
  }

  public static Vector2D operator /(Vector2D v, float s){
    return new Vector2D(v.X / s, v.Y / s);
  }

  public override bool Equals(object obj){
    if(obj is Vector2D v){
      return (this.X == v.X) && (this.Y == v.Y);
    }
    return false;
  }
  
  public override string ToString(){
    return "("+this.X + ", " + this.Y + ")";
  }

};


